crates/sui-graphql-rpc/src/zklogin_verify_signature.rs

```diff
@@ -0,0 +1,199 @@
+    ///    BCS-encoded and then Base64-encoded.
+    /// - `signature` is a serialized zkLogin signature that is Base64-encoded.
+    /// - `intent_scope` is a u64 representing the intent scope of bytes.
+    /// - `author` is the address of the signer of the transaction or personal msg.
+    pub(crate) async fn verify_zklogin_signature(
+        db: &Db,
+        bytes: String,
+        signature: String,
+        intent_scope: u64,
+        author: SuiAddress,
+    ) -> Result<ZkLoginVerifyResult, Error> {
+        // parse intent scope from u64.
+        let safe_intent = if intent_scope > u8::MAX as u64 {
+            return Err(Error::Client("Invalid intent scope".to_string()));
+        } else {
+            IntentScope::try_from(intent_scope as u8)
+                .map_err(|_| Error::Client("Unsupported intent scope".to_string()))
+        }?;
+
+        // get current epoch from db.
+        let Some(curr_epoch) = Epoch::query(db, None, None).await? else {
+            return Err(Error::Internal(
+                "Cannot get current epoch from db".to_string(),
+            ));
+        };
+        let curr_epoch = curr_epoch.stored.epoch as u64;
+
+        // get chain id from db and determine zklogin_env.
+        let chain_id = ChainIdentifier::query(db).await?;
+        let zklogin_env = match chain_id == get_mainnet_chain_identifier()
+            || chain_id == get_testnet_chain_identifier()
+        {
+            true => ZkLoginEnv::Prod,
+            _ => ZkLoginEnv::Test,
+        };
+
+        // fetch on-chain JWKs from dynamic field of system object.
+        let df = DynamicField::query(
+            db,
+            SuiAddress::from_str("0x7")
+                .map_err(|_| Error::Internal("Invalid system object".to_string()))?,
+            None,
+            DynamicFieldName {
+                type_: ExactTypeFilter::from_str("u64")
+                    .map_err(|_| Error::Internal("Invalid df name type".to_string()))?,
+                bcs: Base64::from_str("AQAAAAAAAAA=")
+                    .map_err(|_| Error::Internal("Invalid df name bcs".to_string()))?,
+            },
+            DynamicFieldType::DynamicField,
+            None,
+        )
+        .await
+        .map_err(|e| Error::Internal(e.to_string()))?;
+
+        let binding = df.ok_or(Error::Internal("Cannot find df".to_string()))?;
+
+        let move_object = binding
+            .super_
+            .super_
+            .native_impl()
+            .ok_or(Error::Internal("Cannot pase object from df".to_string()))?
+            .data
+            .try_as_move()
+            .ok_or_else(|| Error::Internal("df not a Move object".to_string()))?;
+
+        let inner = bcs::from_bytes::<Field<u64, AuthenticatorStateInner>>(move_object.contents())
+            .map_err(|err| {
+                Error::Client(format!("Invalid bcs for AuthenticatorStateInner {}", err))
+            })?
+            .value;
```

Comment:
I feel like this obfuscates the fact that this fetch should be infallible (right?) If I encounter any of these errors in our logs, I wouldn't think it would be that serious. I think having a top-level error context of, "We failed to fetch the on-chain JWK(s) from system object 0x7 for reason: {}" would be useful here

Reply:
revised.